### PR TITLE
POC: HTML bindings + Hello World Example

### PR DESCRIPTION
**Pull request by gasi**
- [x] Add Elmish bindings for all HTML elements based on [`purescript-react-basic` codegen][codegen].
- [x] Add ‘Hello World’ example showcasing how to render basic HTML tags using Elmish (without using JS/JSX 😜)
- [x] Set up build pipeline with watcher using [Parcel].
- Simplify project setup by:
    - [x] Removing dependency on yarn (modern `npm` is good enough).
    - [x] Running `bower` as `postinstall` step.

### TODO

- [ ] Determine how to best represent event handlers (currently hard-coded to `JsCallback0`).
- [ ] Determine how to represent `Ref` if needed.
- [ ] Implement helpers for `CSS` type.

[parcel]: https://parceljs.org/
[codegen]: https://github.com/lumihq/purescript-react-basic/tree/master/codegen

### CODE REVIEW

File: NOTICE.txt [hunk 1]
```diff
@@ -0,0 +1,205 @@
+purescript-react-basic’s codegen is used in this library, under the terms of the
+Apache License 2.0:
```
Comment (fsoikin) [inline NOTICE.txt]: We cannot anything under Apache license in MyCV. CC @mohanzhang

Comment (marlier) [reply NOTICE.txt]: That's unusual.  @mohanzhang Can you give more context?

**Comment by gasi** [comment]
> This should be in a separate library.
> First, prevent bloat. Don't pile everything into core that doesn't strictly need to be there.
> More specifically, I was planning to have the library named `purescript-elmish-html`, and module - `Elmish.HTML`.

I was wondering whether this should be separate as well. I see the following trade-off: Having the HTML bindings external would reduce this library’s size but it also wouldn’t make it useful out of the box, i.e. there is no way to actually create an app without either pulling in `purescript-elmish-html` and/or setting up JS/JSX pipeline to create `ReactElement`s. For example, how would you add sample projects to this repo without being able to use basic HTML tags?

I see that Elm follows the ‘separate library’ approach with `elm/html` and was surprised because I thought they would provide a ‘batteries included’ 🔋model at least for a basic HTML app. On the other hand, I presume their aspirations are to go beyond HTML, so that makes sense.

React itself is split into `react` and `react-dom`, but there it’s more about having a generic framework (definition of component, state, props, etc.) in `react` that can also be used in say `react-native` (native for mobile phones) without having DOM/HTML references bundled (`react-dom`).

In our case, we already include `Elmish.React.DOM` since it’s our default rendering target. Therefore, I thought it might make sense to include the basic HTML bindings as well.

Beyond that, I am curious what your thoughts are on how to position this library? At a first glance it’s meant to be a port of Elm to PureScript, which would suggest we should follow the `elm` and `elm/html` separation. On the other hand, one of our main differentiators is that we use ‘just React’ for rendering, unlike the other Elm-like ports: [`purescript-spork`](https://github.com/natefaubion/purescript-spork), [`purescript-oak`](https://github.com/ehrenmurdick/purescript-oak), [`purescript-pux`](https://github.com/alexmingoia/purescript-pux), etc. I believe this is big for two reasons:

1. We can say ‘If you know React, you know exactly how to render your app in _Elmish_’. Big win for onboarding!
2. We use React so you can use any of the hundreds of third-party components. Just write some bindings.

To point (1): A typical React developer would expect to be able to use basic HTML tags out of the box, e.g. with `react` and `react-dom` installed, so not including this would seem like an oversight / unnecessary burden.

> Second, falling back to `div`s is what we need to prevent happening in MyCV.

Agreed, we want to prevent that, but that could be done through convention.

**Comment by fsoikin** [comment]
> there is no way to actually create an app

Yes, there is. `JSX` sidecar file is one of the official rendering modes.

> For example, how would you add sample projects to this repo 
I wouldn't. Sample projects go in a separate repo.

**Comment by fsoikin** [comment]
Most of your points are based on a faulty premise that having two separate libraries is somehow more friction than one library. This is not true. 

For our internal use case, we don't even need the HTML part, so that's a no-op.

For a project that _does_ really want to have basic HTML tags, they can just include `react-elmish-html`, and the core will be pulled in as a dependency.

Additionally, plain HTML tags are only good for proof-of-concept of examples, not for a serious project. For a serious project, one ought to use the styleguide approach - i.e. a library of visuals + imports to PureScript, - similar to our internal use case. So, once again, including HTML tags in the box is just bloat in this case.

And I do plan to emphacise this philosiphy in the docs. Along the lines of: we chose not to include HTML tags in the box on purpose, because you shouldn't really need them, except for examples or such, and if you really _do_ want them, we still provide them as a separate library.

**Comment by gasi** [comment]
Sounds good, I can get on board with that. I see value in having everything in one repo as it lets you keep different parts in sync more easily and also point to embedded examples (which are tested against current version), but I am open to trying the more modular approach 👍 

> Additionally, plain HTML tags are only good for proof-of-concept of examples, not for a serious project. For a serious project, one ought to use the styleguide approach - i.e. a library of visuals + imports to PureScript, - similar to our internal use case. So, once again, including HTML tags in the box is just bloat in this case.

I wouldn’t say it’s just for POCs as some projects, especially with experienced frontend people, like to use HTML elements directly, e.g. to stay semantic (`article`, `header`, `footer`, etc.), and then style them with a CSS framework by applying `className="btn …"`. Also, we are currently building our shared libraries with JSX files, but having access to HTML directly could let us skip JS/JSX/TS/TSX for that 😄 

> And I do plan to emphacise this philosiphy in the docs. Along the lines of: we chose not to include HTML tags in the box on purpose, because you shouldn't really need them, except for examples or such, and if you really _do_ want them, we still provide them as a separate library.

I like that.

> Also: I see none of the elements have any required props. Is that really how they are defined?

Good question. Out of the box. React considers all HTML attributes to be optional. I imagine that’s mainly to avoid having to ship prop types / definitions for all possible tags + attributes.

For example, the HTML standard considers `src` on `img` to be required: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img Might be worth looking into it and lock down the types as we go. Not sure if there is an existing list of mandatory HTML attributes on elements, though.

**Comment by gasi** [comment]
Closed in favor of having separate examples in `purescript-elmish-examples` and HTML bindings in `purescript-elmish-html`:

- https://github.com/collegevine/purescript-elmish-html
- https://github.com/collegevine/purescript-elmish-examples/pull/1

**Pull request closed without merging**